R/derive_merged.R

```diff
@@ -0,0 +1,685 @@
+#'   observations in the additional dataset.
+#'
+#' @inheritParams derive_vars_merged
+#'
+#' @details
+#'
+#'   1. The additional dataset is restricted to the observations matching the
+#'   `filter_add` condition.
+#'
+#'   1. The categorization variable is added to the additional dataset.
+#'
+#'   1. If `order` is specified, for each by group the first or last observation
+#'   (depending on `mode`) is selected.
+#'
+#'   1. The categorization variable is merged to the input dataset.
+#'
+#' @author Stefan Bundfuss
+#'
+#' @keywords derivation adam
+#'
+#' @export
+#'
+#' @examples
+#' library(admiral.test)
+#' library(dplyr, warn.conflicts = FALSE)
+#' data("dm")
+#' data("vs")
+#'
+#' wgt_cat <- function(wgt) {
+#'   case_when(wgt < 50 ~ "low",
+#'             wgt > 90 ~ "high",
+#'             TRUE ~ "normal")
+#' }
+#'
+#' derive_var_merged_cat(
+#'   dm,
+#'   dataset_add = vs,
+#'   by_vars = vars(STUDYID, USUBJID),
+#'   order = vars(VSDTC, VSSEQ),
+#'   filter_add = VSTESTCD == "WEIGHT" & substr(VISIT, 1, 9) == "SCREENING" ,
+#'   new_var = WGTBLCAT,
+#'   source_var = VSSTRESN,
+#'   cat_fun = wgt_cat,
+#'   mode = "last"
+#' ) %>%
+#' select(STUDYID, USUBJID, AGE, AGEU, WGTBLCAT)
+derive_var_merged_cat <- function(dataset,
+                                  dataset_add,
+                                  by_vars,
+                                  order = NULL,
+                                  new_var,
+                                  source_var,
+                                  cat_fun,
+                                  filter_add = NULL,
+                                  mode = NULL,
+                                  missing_value = NA_character_) {
+  new_var <- assert_symbol(enquo(new_var))
+  source_var <- assert_symbol(enquo(source_var))
+  filter_add <- assert_filter_cond(enquo(filter_add), optional = TRUE)
+  assert_data_frame(dataset_add, required_vars = quo_c(by_vars, source_var))
+
+  add_data <- filter_if(dataset_add, filter_add) %>%
+    mutate(!!new_var := cat_fun(!!source_var))
+  derive_vars_merged(
+    dataset,
+    dataset_add = add_data,
+    by_vars = by_vars,
+    order = order,
+    new_vars = vars(!!new_var),
+    match_flag = temp_match_flag,
+    mode = mode
+  ) %>%
+    mutate(!!new_var := if_else(temp_match_flag, !!new_var, missing_value, missing_value)) %>%
+    select(-temp_match_flag)
+}
+
+#' Merge an Existence Flag
+#'
+#' Adds a flag variable to the input dataset which indicates if there exists at
+#' least one observation in another dataset fulfilling a certain condition.
+#'
+#' @param dataset_add Additional dataset
+#'
+#'   The variables specified by the `by_vars` parameter are expected.
+#'
+#' @param by_vars Grouping variables
+#'
+#'   *Permitted Values*: list of variables
+#'
+#' @param new_var New variable
+#'
+#'   The specified variable is added to the input dataset.
+#'
+#' @param condition Condition
```

Comment:
Is that needed? The `filter_add` param should do the filtering?

Reply:
In principle yes, but I wanted to distinguish between no data and condition not fulfilled.

I have added an example (and the `missing_value` parameter).

Reply:
OK, it may be good to add some details for the `condition` parameter, then?

Reply:
Yes, I have added a description. I have also updated the details section as the `missing_value` parameter was not mentioned there.